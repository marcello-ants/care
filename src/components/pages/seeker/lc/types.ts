import { GeneralProviderProfile } from '@/types/seeker';

export interface ProviderProfile extends GeneralProviderProfile {
  numberOfReviews: number;
  qualities?: ProviderQualities;
  services?: ProviderServices;
  firstName: string;
}

export interface ProviderQualities {
  certifiedNursingAssistant?: boolean;
  certifiedRegisterNurse?: boolean;
  certifiedTeacher?: boolean;
  childDevelopmentAssociate?: boolean;
  comfortableWithPets?: boolean;
  cprTrained?: boolean;
  crn?: boolean;
  doesNotSmoke?: boolean;
  doula?: boolean;
  earlyChildDevelopmentCoursework?: boolean;
  earlyChildhoodEducation?: boolean;
  firstAidTraining?: boolean;
  nafccCertified?: boolean;
  ownTransportation?: boolean;
  specialNeedsCare?: boolean;
  trustlineCertifiedCalifornia?: boolean;
}

export interface ProviderServices {
  carpooling?: boolean;
  craftAssistance?: boolean;
  errands?: boolean;
  groceryShopping?: boolean;
  laundryAssistance?: boolean;
  lightHousekeeping?: boolean;
  mealPreparation?: boolean;
  swimmingSupervision?: boolean;
  travel?: boolean;
}
