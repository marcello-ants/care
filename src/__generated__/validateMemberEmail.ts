/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: validateMemberEmail
// ====================================================

export interface validateMemberEmail_validateMemberEmail_errors {
  __typename: "MemberEmailAlreadyRegistered" | "MemberEmailInvalid" | "MemberEmailInvalidFormat" | "MemberEmailInvalidLength";
  /**
   * Summary of the error.
   */
  message: string | null;
}

export interface validateMemberEmail_validateMemberEmail {
  __typename: "ValidateMemberEmailPayload";
  /**
   * Array of email validation errors.  The array will be empty when no validation errors were found.
   */
  errors: validateMemberEmail_validateMemberEmail_errors[];
}

export interface validateMemberEmail {
  /**
   * Validates a member's email address.
   */
  validateMemberEmail: validateMemberEmail_validateMemberEmail;
}

export interface validateMemberEmailVariables {
  email: string;
}
