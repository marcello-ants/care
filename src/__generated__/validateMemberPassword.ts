/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: validateMemberPassword
// ====================================================

export interface validateMemberPassword_validateMemberPassword_errors {
  __typename: "MemberPasswordInvalid" | "MemberPasswordInvalidLength" | "MemberPasswordInvalidSequence" | "MemberPasswordInvalidTerms";
  /**
   * Summary of the error.
   */
  message: string | null;
}

export interface validateMemberPassword_validateMemberPassword {
  __typename: "ValidateMemberPasswordPayload";
  /**
   * Array of password validation errors.  The array will be empty when no validation errors were found.
   */
  errors: validateMemberPassword_validateMemberPassword_errors[];
}

export interface validateMemberPassword {
  /**
   * Validates a member's password.
   */
  validateMemberPassword: validateMemberPassword_validateMemberPassword;
}

export interface validateMemberPasswordVariables {
  password: string;
}
