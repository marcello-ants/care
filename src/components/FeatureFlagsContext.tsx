import { createContext, useContext, ReactNode, useState } from 'react';
import { LDEvaluationDetail } from 'launchdarkly-node-server-sdk';

interface FeatureFlagsProviderProps {
  children: ReactNode;
  flags: FeatureFlags;
}

export type featureFlagsState = {
  flags: FeatureFlags;
};

export type FeatureFlags = Record<string, LDEvaluationDetail | undefined>;

const initialState: featureFlagsState = {
  flags: {},
};

const FeatureFlagsContext = createContext<featureFlagsState>(initialState);

export const FeatureFlagsProvider = ({ children, flags }: FeatureFlagsProviderProps) => {
  const [state] = useState({ flags });
  return <FeatureFlagsContext.Provider value={state}>{children}</FeatureFlagsContext.Provider>;
};

export const useFeatureFlags = () => useContext(FeatureFlagsContext);
