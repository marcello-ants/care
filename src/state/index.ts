/* eslint no-param-reassign: ["error", { "props": true, "ignorePropertyModificationsFor": ["draft"] }] */
import produce, { enableES5 } from 'immer';
import { CURRENT_STATE_VERSION, CURRENT_STATE_MAJOR_VERSION } from '../constants';
import { AppAction, AppState } from '../types/app';

import {
  SeekerAction,
  initialState as seekerInitialState,
  reducer as seekerReducer,
  validate as validateSeeker,
  doesSeekerStateHaveDefaultKeys,
} from './seeker';

import {
  SeekerCCAction,
  initialState as seekerCCInitialState,
  reducer as seekerCCReducer,
  validate as validateSeekerCC,
  doesSeekerCCStateHaveDefaultKeys,
} from './seekerCC';

import {
  ProviderAction,
  initialState as providerInitialState,
  reducer as providerReducer,
} from './provider';

import {
  ProviderCCAction,
  initialState as providerCCInitialState,
  reducer as providerCCReducer,
} from './providerCC';

import {
  SeekerHKAction,
  initialState as seekerHKInitialState,
  reducer as seekerHKReducer,
} from './seekerHK';

import {
  SeekerPCAction,
  initialState as seekerPCInitialState,
  reducer as seekerPCReducer,
} from './seekerPC';

import {
  SeekerTUAction,
  initialState as seekerTUInitialState,
  reducer as seekerTUReducer,
} from './seekerTU';

import { FlowAction, initialState as flowInitialState, reducer as flowReducer } from './flow';

import { LtcgAction, initialState as ltcgInitialState, reducer as ltcgReducer } from './ltcg';

// immer uses Proxy, but if that's not available the ES5 plugin needs to be enabled
/* istanbul ignore if */
if (typeof Proxy === 'undefined') {
  // thanks, IE 11
  enableES5();
}

export const initialAppState: AppState = {
  flow: flowInitialState,
  seeker: seekerInitialState,
  seekerCC: seekerCCInitialState,
  provider: providerInitialState,
  providerCC: providerCCInitialState,
  seekerHK: seekerHKInitialState,
  seekerPC: seekerPCInitialState,
  seekerTU: seekerTUInitialState,
  ltcg: ltcgInitialState,
  version: CURRENT_STATE_VERSION,
};

export const rootReducer = produce((draft: AppState, action: AppAction) => {
  draft.seeker = seekerReducer(draft.seeker, action as SeekerAction);
  draft.provider = providerReducer(draft.provider, action as ProviderAction);
  draft.providerCC = providerCCReducer(draft.providerCC, action as ProviderCCAction);
  draft.flow = flowReducer(draft.flow, action as FlowAction);
  draft.seekerCC = seekerCCReducer(draft.seekerCC, action as SeekerCCAction);
  draft.seekerHK = seekerHKReducer(draft.seekerHK, action as SeekerHKAction);
  draft.seekerPC = seekerPCReducer(draft.seekerPC, action as SeekerPCAction);
  draft.seekerTU = seekerTUReducer(draft.seekerTU, action as SeekerTUAction);
  draft.ltcg = ltcgReducer(draft.ltcg, action as LtcgAction);

  return draft;
});

export function validateState(state: AppState, pathName: string) {
  // TODO: call jobPost validation here instead of in AppState
  return validateSeeker(state, pathName) && validateSeekerCC(state, pathName);
}

export function isStateVersionCompatible(state: AppState) {
  if (state?.version) {
    const versionComponents = state.version.split('.');
    if (versionComponents.length === 3 && versionComponents[0] === CURRENT_STATE_MAJOR_VERSION) {
      return true;
    }
  }
  return false;
}

export function doesStateHaveDefaultKeys(state: AppState) {
  return (
    doesSeekerStateHaveDefaultKeys(state.seeker) && doesSeekerCCStateHaveDefaultKeys(state.seekerCC)
  );
}
