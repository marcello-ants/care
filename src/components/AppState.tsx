import { useRouter } from 'next/router';
import { createContext, useReducer, useContext, ReactNode, useEffect } from 'react';
import { debounce } from 'lodash-es';
import { configureScope, captureException, Severity } from '@sentry/nextjs';
import {
  initialAppState,
  rootReducer,
  isStateVersionCompatible,
  doesStateHaveDefaultKeys,
  validateState,
} from '@/state';
import patchAppState from '@/state/patch';
import { initialState as flowInitialState } from '@/state/flow';
import logger from '@/lib/clientLogger';
import { ClientLogger } from '@care/utils';
import getVertical from '@/utilities/verticalUtils';
import { SessionStorageHelper } from '@/utilities/SessionStorageHelper';
import { trackPageViewInSiftScience } from '@/utilities/siftScienceHelper';
import { AppAction, AppDispatch, AppState } from '@/types/app';
import { StateCheckingResponse } from '@/types/common';
import {
  LOCAL_STORAGE_STATE_KEY,
  CZEN_BASE_PATH,
  FLOWS,
  POST_A_JOB_ROUTES,
  CLIENT_FEATURE_FLAGS,
} from '@/constants';
import { useFeatureFlags } from '@/components/FeatureFlagsContext';
import { AnalyticsHelper } from '@/utilities/analyticsHelper';

type AppStateProviderProps = {
  children: ReactNode;
  initialStateOverride?: AppState;
  czenJSessionId?: string | undefined;
  referrerCookie?: string | undefined | null;
  userHasAccount?: boolean;
  memberUuid?: string;
};

function getStateFromStorageOrDefault(defaultState: AppState) {
  let stateFromStorageJson;
  const sessionStorageAvailable = SessionStorageHelper.isAvailable();

  if (sessionStorageAvailable) {
    stateFromStorageJson = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
  } else {
    logger.warn({ event: 'sessionStorageUnavailable' });
  }

  if (stateFromStorageJson) {
    const stateFromStorage = JSON.parse(stateFromStorageJson) as AppState;
    try {
      const patchedState = patchAppState(stateFromStorage, defaultState);

      if (isStateVersionCompatible(patchedState) && doesStateHaveDefaultKeys(patchedState)) {
        return patchedState;
      }
      logger.warn({ event: 'invalidStateFromLocalStorage', state: stateFromStorageJson });
    } catch (e) {
      captureException(e, { level: Severity.Warning });
    }
  }
  return defaultState;
}

function persistState(state: AppState) {
  SessionStorageHelper.setItem(LOCAL_STORAGE_STATE_KEY, JSON.stringify(state));
}

const debouncePersistState = debounce(persistState, 500, {
  leading: true,
  trailing: true,
  maxWait: 1000,
});

function reducerWrapper(state: AppState, action: AppAction) {
  const nextState = rootReducer(state, action);
  debouncePersistState(nextState);
  return nextState;
}

export function redirectIfInvalidState(state: AppState, pathname: string): StateCheckingResponse {
  if (
    typeof Intl === 'undefined' ||
    typeof Intl.NumberFormat === 'undefined' ||
    typeof Intl.DateTimeFormat === 'undefined'
  ) {
    // TODO: Remove this with SC-214
    logger.error({ event: 'IntlUnavailable', source: pathname, invalidState: state });
    return {
      redirectCalled: true,
      readableMessage: 'Safari 9 and lower is not currently supported',
    };
  }

  const isStateValid = validateState(state, pathname);
  const redirect = !isStateValid;

  if (redirect) {
    SessionStorageHelper.removeItem(LOCAL_STORAGE_STATE_KEY);

    window.location.assign(CZEN_BASE_PATH);
  }

  return { redirectCalled: redirect };
}

export function getFlowName(pathname: string, oldFlowName?: string): string {
  let newFlowName = Object.values(FLOWS).find((flow) => pathname.includes(flow.path))?.name;

  if (!newFlowName) {
    // TODO: this shouldn't be needed after SC-123 is implemented
    const isPostAJob = Object.values(POST_A_JOB_ROUTES).some((route) => pathname.includes(route));
    if (isPostAJob) {
      newFlowName = 'SEEKER_POST_A_JOB';
    }
  }

  // Don't allow shifting from in-facility to seeker flow name
  if (newFlowName === FLOWS.SEEKER.name && oldFlowName === FLOWS.SEEKER_IN_FACILITY.name) {
    return FLOWS.SEEKER_IN_FACILITY.name;
  }

  return newFlowName ?? '';
}

function useFlowDetection(dispatch: AppDispatch, oldFlowName: string) {
  const { pathname } = useRouter();
  const newFlowName = getFlowName(pathname, oldFlowName);
  const vertical = getVertical();
  ClientLogger.setLoggerStatics({ flow: newFlowName, vertical });

  useEffect(() => {
    dispatch({ type: 'setFlowName', flowName: newFlowName });

    configureScope((scope) => {
      scope.setTags({
        flow: newFlowName,
        vertical,
      });
    });
  }, [newFlowName]);
}

const AppStateContext = createContext<AppState>(initialAppState);
const AppDispatchContext = createContext<AppDispatch>(() =>
  // eslint-disable-next-line no-console
  console.warn(
    'Calling default dispatch is a no-op! You are probably consuming app dispatch in a component above AppStateProvider'
  )
);

export const AppStateProvider = ({
  children,
  initialStateOverride,
  czenJSessionId,
  referrerCookie,
  userHasAccount,
  memberUuid,
}: AppStateProviderProps) => {
  const hydratedInitialAppState = {
    ...initialAppState,
    flow: {
      ...flowInitialState,
      czenJSessionId,
      referrerCookie,
      userHasAccount,
      memberUuid,
    },
  };
  const featureFlags = useFeatureFlags();
  const sessionStorageRetrievalSuccessEvaluation =
    featureFlags?.flags[CLIENT_FEATURE_FLAGS.SESSION_STORAGE_RETRIEVAL_SUCCESS];
  const shouldMeasureSessionStorageRetrievalSuccess =
    sessionStorageRetrievalSuccessEvaluation?.variationIndex === 1;

  const router = useRouter();

  const logSessionStorageError = () => {
    logger.info({ event: 'sessionStorageStateUnavailable', path: router.pathname });
    AnalyticsHelper.logEvent({
      name: 'sessionStorageStateUnavailable',
    });
  };
  const launchDarklyWrappedGetStateFromStorageOrDefault = (defaultState: AppState) => {
    if (shouldMeasureSessionStorageRetrievalSuccess) {
      try {
        const sessionStorageState = SessionStorageHelper.getItem(LOCAL_STORAGE_STATE_KEY);
        if (sessionStorageState && sessionStorageState.length > 0) {
          const stateFromStorage = JSON.parse(sessionStorageState);
          if (stateFromStorage) {
            logger.info({ event: 'sessionStorageStateAvailable', path: router.pathname });
          } else {
            logSessionStorageError();
          }
        } else {
          logSessionStorageError();
        }
      } catch (_e) {
        logSessionStorageError();
      }
    }
    return getStateFromStorageOrDefault(defaultState);
  };

  const [state, dispatch] = useReducer(
    reducerWrapper,
    initialStateOverride ?? hydratedInitialAppState,
    launchDarklyWrappedGetStateFromStorageOrDefault
  );
  useFlowDetection(dispatch, state.flow.flowName);
  const routePath = router.asPath?.split('?')[0];
  const stateCheckResult = redirectIfInvalidState(state, routePath);
  if (stateCheckResult.redirectCalled) {
    if (stateCheckResult.readableMessage) {
      return (
        <>
          <p>Redirecting to Member Home Page...</p>
          <p>{stateCheckResult.readableMessage}</p>
        </>
      );
    }
    return <></>;
  }

  useEffect(() => {
    trackPageViewInSiftScience(state);
  }, [routePath]);

  return (
    <AppDispatchContext.Provider value={dispatch}>
      <AppStateContext.Provider value={state}>{children}</AppStateContext.Provider>
    </AppDispatchContext.Provider>
  );
};

AppStateProvider.defaultProps = {
  initialStateOverride: undefined,
  czenJSessionId: undefined,
  referrerCookie: undefined,
  userHasAccount: false,
  memberUuid: undefined,
};

export const useAppState = () => useContext(AppStateContext);
export const useAppDispatch = () => useContext(AppDispatchContext);

export const useSeekerState = () => useContext(AppStateContext).seeker;
export const useProviderState = () => useContext(AppStateContext).provider;
export const useProviderCCState = () => useContext(AppStateContext).providerCC;
export const useFlowState = () => useContext(AppStateContext).flow;
export const useSeekerCCState = () => useContext(AppStateContext).seekerCC;
export const useSeekerHKState = () => useContext(AppStateContext).seekerHK;
export const useSeekerPCState = () => useContext(AppStateContext).seekerPC;
export const useSeekerTUState = () => useContext(AppStateContext).seekerTU;
export const useLtcgState = () => useContext(AppStateContext).ltcg;
