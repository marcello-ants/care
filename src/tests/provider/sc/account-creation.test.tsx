import React from 'react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { useRouter } from 'next/router';
import { ThemeProvider } from '@material-ui/core';
import { theme } from '@care/material-ui-theme';

import {
  GET_AVERAGE_JOB_WAGE,
  GET_NUMBER_OF_JOBS_NEARBY,
  VALIDATE_MEMBER_EMAIL,
  VALIDATE_MEMBER_PASSWORD,
  PROVIDER_CREATE,
} from '@/components/request/GQL';
import { ServiceType, DistanceUnit } from '@/__generated__/globalTypes';
// eslint-disable-next-line camelcase
import { validateMemberEmail_validateMemberEmail_errors } from '@/__generated__/validateMemberEmail';
// eslint-disable-next-line camelcase
import { validateMemberPassword_validateMemberPassword_errors } from '@/__generated__/validateMemberPassword';
import AccountCreation from '@/pages/provider/sc/account-creation/[[...param]]';
import { AppStateProvider } from '@/components/AppState';
import { FeatureFlagsProvider } from '@/components/FeatureFlagsContext';
import {
  CLIENT_FEATURE_FLAGS,
  EMAIL_PASSWORD_NAME_JOINT_VALIDATION_ERROR,
  PROVIDER_ROUTES,
} from '@/constants';
import { initialAppState } from '@/state';
import { AppState } from '@/types/app';
import AuthService from '@/lib/AuthService';

jest.mock('next/router', () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));

jest.mock('next/config', () => () => {
  return {
    publicRuntimeConfig: {
      OIDC_STS_AUTHORITY: 'OIDC_STS_AUTHORITY',
      OIDC_CLIENT_ID: 'OIDC_CLIENT_ID',
      OIDC_RESPONSE_TYPE: 'OIDC_RESPONSE_TYPE',
      OIDC_CLIENT_SCOPE: 'OIDC_CLIENT_SCOPE',
      OIDC_CALLBACK_PATH: 'OIDC_CALLBACK_PATH',
      OIDC_LOGOUT_URL: 'OIDC_LOGOUT_URL',
    },
  };
});

jest.mock('@/lib/AuthService');
const AuthServiceMock = AuthService as jest.Mock;

const redirectLoginSpy = jest.fn();
AuthServiceMock.mockImplementation(() => {
  return {
    redirectLogin: redirectLoginSpy,
  };
});

describe('AccountCreation', () => {
  let mockRouter: any | null = null;
  let emailField: HTMLElement;
  let passwordField: HTMLElement;
  let joinNowBtn: HTMLElement;

  const initialState: AppState = {
    ...initialAppState,
    provider: {
      ...initialAppState.provider,
      zipcode: '12345',
      distance: 25,
    },
  };

  const nearbyJobsMock = {
    request: {
      query: GET_NUMBER_OF_JOBS_NEARBY,
      variables: {
        zipcode: '12345',
        serviceType: 'SENIOR_CARE',
      },
    },
    result: {
      data: {
        getNumberOfNewJobsNearby: 50,
      },
    },
  };

  const wageMock = {
    request: {
      query: GET_AVERAGE_JOB_WAGE,
      variables: {
        zipcode: '12345',
        serviceType: 'SENIOR_CARE',
      },
    },
    result: {
      data: {
        getJobWages: { averages: { average: { amount: 15 } } },
      },
    },
  };

  function emailValidationMock(
    email: string,
    // eslint-disable-next-line camelcase
    error?: validateMemberEmail_validateMemberEmail_errors
  ) {
    return {
      request: {
        query: VALIDATE_MEMBER_EMAIL,
        variables: {
          email,
        },
      },
      result: {
        data: {
          validateMemberEmail: { errors: [error] },
        },
      },
    };
  }

  function passwordValidationMock(
    password: string,
    // eslint-disable-next-line camelcase
    error?: validateMemberPassword_validateMemberPassword_errors
  ) {
    return {
      request: {
        query: VALIDATE_MEMBER_PASSWORD,
        variables: {
          password,
        },
      },
      result: {
        data: {
          validateMemberPassword: { errors: [error] },
        },
      },
    };
  }

  const failingAccountCreateMsg = 'An error occurred creating your account, please try again.';
  const providerCreateFailureMock = {
    request: {
      query: PROVIDER_CREATE,
      variables: {
        input: {
          serviceType: ServiceType.SENIOR_CARE,
          email: 'tim@care.com',
          howDidYouHearAboutUs: undefined,
          distanceWillingToTravel: { value: 25, unit: DistanceUnit.MILES },
          password: 'test456',
          referrerCookie: initialState.flow.referrerCookie,
          zipcode: '12345',
        },
      },
    },
    result: {
      data: {
        providerCreate: {
          errors: [
            {
              __typename: 'ProviderCreateError',
              message: failingAccountCreateMsg,
            },
          ],
          __typename: 'ProviderCreateError',
        },
      },
    },
  };

  const providerCreateSuccessMock = {
    request: {
      query: PROVIDER_CREATE,
      variables: {
        input: {
          serviceType: ServiceType.SENIOR_CARE,
          email: 'tim@care.com',
          howDidYouHearAboutUs: undefined,
          distanceWillingToTravel: { value: 25, unit: DistanceUnit.MILES },
          password: 'test456',
          referrerCookie: initialState.flow.referrerCookie,
          zipcode: '12345',
        },
      },
    },
    result: {
      data: {
        providerCreate: {
          __typename: 'ProviderCreateSuccess',
          authToken: 'pPMAzhsSOJx5ZxZAy_e27*n28NNwBug4D0III1hM1Dw.',
          memberId: '4747',
        },
      },
    },
  };

  const featureFlagsMockTrue = {
    [CLIENT_FEATURE_FLAGS.PROVIDER_CC_FREE_GATED_EXPERIENCE]: {
      value: true,
      reason: {
        kind: 'FALLTHROUGH',
      },
    },
  };

  function renderPage(mocks: ReadonlyArray<MockedResponse> = []) {
    const view = render(
      <MockedProvider mocks={[nearbyJobsMock, wageMock, ...mocks]} addTypename={false}>
        <ThemeProvider theme={theme}>
          <AppStateProvider initialStateOverride={initialState}>
            <FeatureFlagsProvider flags={featureFlagsMockTrue}>
              <AccountCreation />
            </FeatureFlagsProvider>
          </AppStateProvider>
        </ThemeProvider>
      </MockedProvider>
    );

    emailField = screen.getByLabelText('Email address');
    passwordField = screen.getByLabelText('Password');
    joinNowBtn = screen.getByRole('button', { name: 'Join now' });

    return view;
  }

  beforeEach(() => {
    mockRouter = {
      push: jest.fn(),
      replace: jest.fn(),
      query: { param: [] },
      pathname: '',
    };
    (useRouter as jest.Mock).mockReturnValue(mockRouter);
  });

  afterEach(() => {
    // cleanup on exiting
    mockRouter = null;
  });

  it('matches snapshot', () => {
    const { asFragment } = renderPage();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly', () => {
    renderPage();
    expect(screen.getByText('Create an account to apply for jobs')).toBeInTheDocument();
  });

  it('should initially render the submit button as disabled', async () => {
    renderPage();

    await waitFor(() => expect(joinNowBtn).toBeDisabled());
  });

  it('should disable the submit button if an invalid email is entered', async () => {
    renderPage([
      emailValidationMock('tim@care.com'),
      emailValidationMock('invalid@care.com', {
        __typename: 'MemberEmailInvalid',
        message: 'bad email',
      }),
      passwordValidationMock('test456'),
    ]);

    fireEvent.change(emailField, { target: { value: 'tim@care.com' } });
    fireEvent.change(passwordField, { target: { value: 'test456' } });

    await waitFor(() => expect(joinNowBtn).toBeEnabled());

    fireEvent.change(emailField, { target: { value: 'invalid@care.com' } });

    await waitFor(() => expect(joinNowBtn).toBeDisabled());
  });

  it('should disable the submit button if an invalid password is entered', async () => {
    renderPage([
      emailValidationMock('tim@care.com'),
      passwordValidationMock('test456'),
      passwordValidationMock('password', {
        __typename: 'MemberPasswordInvalidTerms',
        message: 'nope',
      }),
    ]);

    fireEvent.change(emailField, { target: { value: 'tim@care.com' } });
    fireEvent.change(passwordField, { target: { value: 'test456' } });

    await waitFor(() => expect(joinNowBtn).toBeEnabled());

    fireEvent.change(passwordField, { target: { value: 'password' } });

    await waitFor(() => expect(joinNowBtn).toBeDisabled());
  });

  it('should disable the submit button when the email is in the password', async () => {
    renderPage([emailValidationMock('tim@care.com'), passwordValidationMock('test456')]);

    fireEvent.change(emailField, { target: { value: 'tim@care.com' } });
    fireEvent.change(passwordField, { target: { value: 'tim@care.com' } });

    await waitFor(() => expect(joinNowBtn).toBeDisabled());
    expect(await screen.findByText(EMAIL_PASSWORD_NAME_JOINT_VALIDATION_ERROR)).toBeVisible();
  });

  it('should redirect via AuthService then route to the account created page on successful submission', async () => {
    renderPage([
      emailValidationMock('tim@care.com'),
      passwordValidationMock('test456'),
      providerCreateSuccessMock,
    ]);

    fireEvent.change(emailField, { target: { value: 'tim@care.com' } });
    fireEvent.change(passwordField, { target: { value: 'test456' } });

    await waitFor(() => expect(joinNowBtn).toBeEnabled());

    fireEvent.click(joinNowBtn);
    // AuthService.redirectLogin() will only be called in the case where our
    // mutation mock was found and invoked
    await waitFor(() => expect(redirectLoginSpy).toHaveBeenCalledTimes(1));
  });

  it('Checks the error message in the screen when mutation for account creation fails', async () => {
    renderPage([
      emailValidationMock('tim@care.com'),
      passwordValidationMock('test456'),
      providerCreateFailureMock,
    ]);

    fireEvent.change(emailField, { target: { value: 'tim@care.com' } });
    fireEvent.change(passwordField, { target: { value: 'test456' } });

    await waitFor(() => expect(joinNowBtn).toBeEnabled());

    fireEvent.click(joinNowBtn);

    expect(await screen.findByText(failingAccountCreateMsg)).toBeVisible();
  });

  it('should redirect back to the index page when the zipcode is missing', async () => {
    const initialStateOverride: AppState = {
      ...initialState,
      provider: {
        ...initialState.provider,
        zipcode: '',
      },
    };

    render(
      <MockedProvider mocks={[nearbyJobsMock, wageMock]} addTypename={false}>
        <ThemeProvider theme={theme}>
          <AppStateProvider initialStateOverride={initialStateOverride}>
            <FeatureFlagsProvider flags={featureFlagsMockTrue}>
              <AccountCreation />
            </FeatureFlagsProvider>
          </AppStateProvider>
        </ThemeProvider>
      </MockedProvider>
    );

    await waitFor(() => expect(mockRouter.replace).toBeCalledWith(PROVIDER_ROUTES.INDEX));
  });
});
