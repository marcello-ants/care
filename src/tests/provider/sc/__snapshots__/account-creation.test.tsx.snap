// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccountCreation matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root makeStyles-gridContainer-1 MuiGrid-container"
  >
    <div
      class="MuiGrid-root makeStyles-alertContainer-2 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiBox-root MuiBox-root-11"
      >
        <div
          class="MuiPaper-root MuiAlert-root MuiAlert-standardInfo makeStyles-alertContent-9 MuiPaper-elevation0"
          role="alert"
        >
          <div
            class="MuiAlert-icon"
          >
            <p
              class="MuiTypography-root makeStyles-alertEmoji-10 MuiTypography-body2"
            >
              <span>
                ðŸŽ‰
              </span>
            </p>
          </div>
          <div
            class="MuiAlert-message"
          >
            <p
              class="MuiTypography-root MuiTypography-body2"
            >
              <span>
                <span>
                  There are new jobs near you.
                </span>
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <h2
        class="MuiTypography-root makeStyles-header-26 MuiTypography-h2"
      >
        <span>
          Create an account to apply for jobs
        </span>
      </h2>
    </div>
    <div
      class="MuiGrid-root makeStyles-inputContainer-5 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-inline-27"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="email"
          id="email-label"
        >
          Email address
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="email"
            name="email"
            type="email"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root makeStyles-inputContainer-5 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-inline-27"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Password
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
        >
          <input
            aria-describedby="password-helper-text"
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <p
          class="MuiFormHelperText-root"
          id="password-helper-text"
        >
          At least 6 characters
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-root makeStyles-inputContainer-5 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <div
        class="MuiFormControl-root makeStyles-selectContainer-31"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
          data-shrink="false"
          id="source"
        >
          How did you hear about us?
        </label>
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline no-value MuiInputBase-formControl MuiInput-formControl"
        >
          <div
            aria-haspopup="listbox"
            aria-labelledby="source"
            class="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
            id="source"
            role="button"
            tabindex="-1"
          >
            Select option
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput"
            name="source"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-root-36"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M18.2 8a.91.91 0 011.2 0 .91.91 0 010 1.2L12.6 16a.91.91 0 01-1.2 0L4.6 9.2a.91.91 0 010-1.2.91.91 0 011.2 0l6.2 6.2z"
              fill="#223f5f"
            />
            <path
              d="M12 16.37a1 1 0 01-.71-.26l-6.8-6.8a1.09 1.09 0 010-1.42 1.09 1.09 0 011.42 0L12 14l6.09-6.1a1.09 1.09 0 011.42 0 1.09 1.09 0 010 1.42l-6.8 6.8a1 1 0 01-.71.25zM5.2 7.93a.69.69 0 00-.49.18.76.76 0 000 1l6.8 6.8a.78.78 0 001 0l6.8-6.8a.76.76 0 000-1 .76.76 0 00-1 0L12 14.41l-.11-.1-6.2-6.2a.69.69 0 00-.49-.18z"
              fill="#223f5f"
            />
          </svg>
        </div>
        <p
          class="MuiFormHelperText-root"
        >
          <span>
            â€‹
          </span>
        </p>
      </div>
    </div>
    <div
      class="MuiGrid-root makeStyles-legalTextContainer-3 MuiGrid-item MuiGrid-grid-xs-12"
    >
      <p
        class="MuiTypography-root makeStyles-legalText-4 MuiTypography-body2"
      >
        <span>
          By clicking "Join now", you agree to our
        </span>
         
        <a
          href="https://www.care.com/terms-of-use-popup-p1030.html"
          rel="noreferrer"
          tabindex="-1"
          target="_blank"
        >
          Terms of Use
        </a>
         
        <span>
          and
        </span>
         
        <a
          href="https://www.care.com/privacy-policy-popup-p1031.html"
          rel="noreferrer"
          tabindex="-1"
          target="_blank"
        >
          Privacy Policy
        </a>
        <span>
          .
        </span>
      </p>
    </div>
    <div
      class="makeStyles-wrapper-34"
    >
      <div
        class="makeStyles-fade-35"
        style="position: relative; width: auto; margin-bottom: -24px; padding: 32px 40px 24px;"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Join now
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
